src/extract_gpuinfo_ascend.c

```diff
@@ -0,0 +1,251 @@
+  struct list_head allocate_list;
+};
+
+static LIST_HEAD(allocations);
+
+static bool gpuinfo_ascend_init(void);
+static void gpuinfo_ascend_shutdown(void);
+static const char *gpuinfo_ascend_last_error_string(void);
+static bool gpuinfo_ascend_get_device_handles(struct list_head *devices, unsigned *count);
+static void gpuinfo_ascend_populate_static_info(struct gpu_info *_gpu_info);
+static void gpuinfo_ascend_refresh_dynamic_info(struct gpu_info *_gpu_info);
+static void gpuinfo_ascend_get_running_processes(struct gpu_info *_gpu_info);
+
+static void _encode_card_device_id_to_pdev(char *pdev, int card_id, int device_id);
+static void _decode_card_device_id_from_pdev(const char *pdev, int *card_id, int *device_id);
+
+struct gpu_vendor gpu_vendor_ascend = {
+    .init = gpuinfo_ascend_init,
+    .shutdown = gpuinfo_ascend_shutdown,
+    .last_error_string = gpuinfo_ascend_last_error_string,
+    .get_device_handles = gpuinfo_ascend_get_device_handles,
+    .populate_static_info = gpuinfo_ascend_populate_static_info,
+    .refresh_dynamic_info = gpuinfo_ascend_refresh_dynamic_info,
+    .refresh_running_processes = gpuinfo_ascend_get_running_processes,
+    .name = "Ascend",
+};
+
+__attribute__((constructor)) static void init_extract_gpuinfo_ascend(void) { register_gpu_vendor(&gpu_vendor_ascend); }
+
+static bool gpuinfo_ascend_init(void) {
+  last_dcmi_return_status = dcmi_init();
+  return last_dcmi_return_status == DCMI_SUCCESS;
+}
+
+static void gpuinfo_ascend_shutdown(void) {
+  local_error_string = "";
+
+  struct gpu_info_ascend *allocated, *tmp;
+  list_for_each_entry_safe(allocated, tmp, &allocations, allocate_list) {
+    list_del(&allocated->allocate_list);
+    free(allocated);
+  }
+}
+
+static const char *gpuinfo_ascend_last_error_string(void) {
+  return local_error_string;
+}
+
+static bool gpuinfo_ascend_get_device_handles(struct list_head *devices, unsigned *count) {
+  int num_cards;
+  int card_list[MAX_CARD_NUM] = {0};
+  last_dcmi_return_status = dcmi_get_card_list(&num_cards, card_list, MAX_DEVICE_NUM);
+  if (last_dcmi_return_status != DCMI_SUCCESS) {
+    local_error_string = "Failed to get card num";
+    return false;
+  } else if (num_cards == 0) {
+    local_error_string = "Not found NPU(s)";
+    return false;
+  }
+
+  int num_devices = 0;
+  int card_device_list[num_cards];
+  for (int i = 0; i < num_cards; ++i) {
+    int num_card_devices;
+    last_dcmi_return_status = dcmi_get_device_num_in_card(card_list[i], &num_card_devices);
+    if (last_dcmi_return_status != DCMI_SUCCESS) {
+      local_error_string = "Failed to get device num of card";
+      return false;
+    }
+    num_devices += num_card_devices;
+    card_device_list[i] = num_card_devices;
+  }
+
+  struct gpu_info_ascend *gpu_infos = calloc(num_devices, sizeof(*gpu_infos));
+  if (!gpu_infos) {
+    local_error_string = strerror(errno);
+    return false;
+  }
+
+  // todo: for free gpu_infos when shutting down, rewrite to direct free?
```

Comment:
I see the problem, I should add the device list to the shutdown function for a proper way to do the cleanup.
For now it's good as is!

Reply:
Okay